Automattic/wp-calypso
# Checkout: Accept redirect_to qarg, redirect there post-purchase

#### Changes proposed in this Pull Request

* Allow appending a `redirect_to` query arg to `/checkout` routes, and redirect to the specified URL post checkout.

#### Implementation Notes

I was delighted to find that `my-sites/checkout/checkout/` already supported a `redirectTo` mechanism and was able to piggy-back on it. However, it only accepts server-relative URLs, as it's appended to the current host and port: https://github.com/Automattic/wp-calypso/blob/8f7a7f3ab8c045e9d7d35ae84e08ba5141e8d633/client/my-sites/checkout/checkout/secure-payment-form.jsx#L167
On the plus side, this mitigates the validation question -- the worst that a malicious actor can do is send the user to a different route on WordPress.com (but nowhere external).

For out intended purpose (coming from the block editor through and upgrade nudge on a WP.com site), this restriction shouldn't pose a problem. However, it means we'll have to revisit this once we want to make this available to Jetpack sites.

Furthermore, note that this adds yet another way to customize the checkout flow, adding to a number of existing ones. I feel slightly less bad since the `redirect_to` qarg is fairly explicit and visible, piggy-backs onto an existing mechanism (`redirectTo` in `checkout/checkout/index.jsx`), and has a small enough impact surface (just passing the query arg as a prop through the component hierarchy). 

#### Testing instructions

* For any WP.com `:site` of yours (ideally on a Free Plan), navigate to `http://calypso.localhost:3000/checkout/:site/premium?redirect_to=/post/:site/3`
* Complete chekout
* Verify that upon completion, you are indeed redirected to `http://calypso.localhost:3000/post/:site/3`

#### Design Questions

This PR includes a number of implicit design decisions, which need approval or advice by a designer, specifically about the the exit point for the checkout flow: Specifically, I chose to by-pass the 'Thank You' page, which I thought would be too distracting and disruptive to the flow -- assuming that we want to get the user to get the task done that they're after: upgrade to the required plan as swiftly as possible.

This doesn't preclude e.g. adding a less intrusive 'Thank You' notice to the editor, once they've been redirected back (we did something similar in the theme purchase flow). Due to the nature of our blocks living in the Jetpack repo, we would have to implement that notice there, though.

Anyway, curious to hear a designer's POV.

~(A similar question applies to the entry-point of the flow. Right now, this PR is assuming that the upgrade nudge takes the user to `/plans/:site?redirect_to=/:site?redirect_to=/post/:site/:postId`, allowing them to select a plan. This gives the user a degree of freedom: They can opt to buy the suggested (minimum required) plan, or a higher tier one (more $$$ for us), or a lower tier one -- in which case the block won't work! An alternative to consider is thus to have the upgrade nudge take them directly to `/checkout/:site/:plan?redirect_to=/post/:site/:postId`, removing that degree of freedom. I'm leaning towards the latter, but this is something we will need to implement on the Jetpack side, so it's not _that_ relevant for this PR. Just mentioning for completeness' sake.)~

Edit: Changed to only touch `/checkout` but not `/plans` routes (per https://github.com/Automattic/wp-calypso/pull/34440#issuecomment-508386607)

Fixes #34400
